### PR TITLE
[TG-175][Dima D] Update Plug-n-Play logic to optimize hardware interaction

### PR DESCRIPTION
**Pull request by demakoff**

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/garage-it/SmartHouse-backend/pull/34" target="_blank">ReviewNinja

File: src/mqtt-client/client.js [hunk 1]
```diff
@@ -36,6 +39,16 @@ function onConnect() {
         } else  if (event.event === 'device-info') {
```
Comment (antonSavk) [inline src/mqtt-client/client.js]: Consider moving event name to a constant

File: src/mqtt-client/client.js [hunk 2]
```diff
@@ -59,6 +72,16 @@ function onSubscribed() {
             deviceName = topic.split('/').pop();
 
         if (isDeviceInfo) {
+            debug(`Device-info came: ${deviceName}, publishingQueue.size: ${publishingQueue.size}`);
+            publishingQueue.delete(deviceName);
+
+            if (publishingQueue.size) {
+                debug(`Publish from queue: '${[...publishingQueue][0]}'`);
+                client.publish(MQTT_OUTPUT_TOPIC_PREFIX + 'device-info', [...publishingQueue][0]);
```
Comment (antonSavk) [inline src/mqtt-client/client.js]: Consider using string literals for strings concatenation

File: src/mqtt-client/client.js [hunk 3]
```diff
@@ -59,6 +72,16 @@ function onSubscribed() {
...
+            } else {
+                blockPublishing = false;
+            }
+
             event  = {
                 event: 'device-info',
```
Comment (antonSavk) [inline src/mqtt-client/client.js]: same here

File: src/config/env/cloud.js [hunk 4]
```diff
@@ -3,7 +3,7 @@ export default {
     db: 'mongodb://localhost/db',
     port: 3000,
     seedDB: false,
-    plugAndPlay: false,
+    plugAndPlay: true,
```
Comment (kosich) [inline src/config/env/cloud.js]: !fix
lets keep that disabled for cloud -- which would be used by all developers -- which can lead to ddosing device by local servers
and enable that for heroku env

File: src/mqtt-client/client.js [hunk 5]
```diff
@@ -22,6 +22,9 @@ const client = mqtt.connect({
     auth: `${config.mqtt.username}:${config.mqtt.password}`
 });
 
+var blockPublishing = false,
+    publishingQueue = new Set();
```
Comment (kosich) [inline src/mqtt-client/client.js]: use `let` or `const`
and each new definition with its own `let` or `const`

File: src/mqtt-client/client.js (hunk 2 above)
Comment (kosich) [inline src/mqtt-client/client.js]: wont be necessary if `'device-info'` moved to a constant

File: src/mqtt-client/client.js [hunk 6]
```diff
@@ -59,6 +72,16 @@ function onSubscribed() {
             deviceName = topic.split('/').pop();
 
         if (isDeviceInfo) {
+            debug(`Device-info came: ${deviceName}, publishingQueue.size: ${publishingQueue.size}`);
```
Comment (kosich) [inline src/mqtt-client/client.js]: !fix
if we don't get this message (e.g. when device went off) -- we wont proceed with other requests -- add timeout for discarding request (and moving that to the end of queue)

also consider moving logic with plug-n-play to a separate file

File: src/mqtt-client/client.js (hunk 1 above)
Comment (demakoff) [inline src/mqtt-client/client.js]: ok

File: src/mqtt-client/client.js (hunk 3 above)
Comment (demakoff) [inline src/mqtt-client/client.js]: I'll leave it for a refactoring

File: src/config/env/cloud.js (hunk 4 above)
Comment (demakoff) [inline src/config/env/cloud.js]: ok

File: src/mqtt-client/client.js (hunk 5 above)
Comment (demakoff) [inline src/mqtt-client/client.js]: ok

File: src/mqtt-client/client.js (hunk 6 above)
Comment (demakoff) [inline src/mqtt-client/client.js]: Regarding timeout I agree. About extract to separate file - I thought a lot about it but I will leave it for refactoring also...

File: src/config/env/cloud.js (hunk 4 above)
Comment (demakoff) [inline src/config/env/cloud.js]: !fixed

File: src/mqtt-client/client.js (hunk 6 above)
Comment (demakoff) [inline src/mqtt-client/client.js]: !fixed

File: src/mqtt-client/client.js [hunk 7]
```diff
@@ -77,4 +90,28 @@ function onSubscribed() {
...
+        publishingQueue.add(device);
+    }
+
+    if (!blockPublishing) {
+        if (timeoutTimer) {
+            clearTimeout(timeoutTimer);
+        }
+        if (publishingQueue.size) {
+            blockPublishing = true;
+            client.publish(MQTT_OUTPUT_TOPIC_PREFIX + 'device-info', [...publishingQueue][0]);
```
Comment (kosich) [inline src/mqtt-client/client.js]: use `DEVICE_INFO_EVENT`

**Comment by kosich** [comment]
⭐

**Comment by kosich** [comment]
⭐

**Pull request merged**